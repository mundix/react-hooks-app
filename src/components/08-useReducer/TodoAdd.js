import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    //Creo el submit aqui
    const handleSubmit = (e) => {
        
        e.preventDefault();
        
        if (description.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        // dispatch(action); //Este no es el dispach , con el handleADdtodo en ves del dispatch
        // Esta funcion se llama desde el padre del component que recibe un newTodo, y se encarga de hacer el dispatch
        handleAddTodo(newTodo);
        reset();
    }

    return (
        // La forma corta del Fragment asi se llama este elemento <> que es un root element del component 
        <>
            <h4>Agregar Todo</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    onChange={handleInputChange}
                    value={description}
                    type="text"
                    name='description'
                    className='form-control'
                    placeholder='Aprender ...'
                    autoComplete='off'
                />
                <button
                    className='btn btn-outline-warning mt-1 btn-block'
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
