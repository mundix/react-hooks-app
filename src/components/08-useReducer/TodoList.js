import React from 'react';

export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
        <>
            <ul className='list-group list-group-flush'>
                        {
                            todos.map((todo, i) => (
                                <li
                                    key={todo.id}
                                    className='list-group-item'
                                >
                                    <p
                                        className={`${todo.done && 'complete'}`}
                                        onClick={() => handleToggle(todo.id)}

                                    > {i + 1} - {todo.desc}</p>
                                    <button
                                        className="btn btn-danger"
                                        // Como necesito pasar argumento , se envia un callback y la funciona con el argumento
                                        onClick={() => handleDelete(todo.id)}
                                    >
                                        Borrad
                                    </button>
                                </li>
                            ))
                        }
                    </ul>
        </>
    )
}
