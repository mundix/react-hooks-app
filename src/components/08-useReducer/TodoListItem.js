import React from 'react'

export const TodoListItem = ({ todo, index, handleDelete, handleToggle }) => {
    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p
                className={`${todo.done && 'complete'}`}
                onClick={() => handleToggle(todo.id)}

            > {index + 1} - {todo.desc}</p>
            <button
                className="btn btn-danger"
                // Como necesito pasar argumento , se envia un callback y la funciona con el argumento
                onClick={() => handleDelete(todo.id)}
            >
                Borrad
            </button>
        </li>
    )
}
