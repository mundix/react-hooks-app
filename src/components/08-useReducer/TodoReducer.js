export const todoReducer = (state = [], action) => {
    switch (action.type) {
        
        case 'add':
            return [...state, action.payload];
        
        case 'delete':
            return state.filter(todo => todo.id !== action.payload);
        
        case 'toggle' : //Version mas corta que el old
            return state.map( todo => 
                    (todo.id === action.payload) 
                    ? {...todo, done: !todo.done} 
                    : todo
                )
        case 'toggle-old':
            return state.map(todo => {
                // Asumiento en el payload.id viene en el mismo 
                if (todo.id === action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    // Debe retonar algo para no mutar y destruir algo , porque si no va a trasnforma en undefine 
                    return todo;
                }
            }); //hay que devovler un nuego state

        default:
            return state;
    }
}