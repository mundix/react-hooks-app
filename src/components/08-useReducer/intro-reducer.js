const initialState = [{  //se puede poenr cualquier nombre de variable, pero este es el valor inicial
    id: 1,
    todo: 'Comprar Pan',
    done: false
}];

// Se crea una function 
const todoReducer = ( state = initialState , action) => {

    if(action?.type === 'agregar') {
        return [ ...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2, 
    todo: 'Comprar Leche',
    done: false
}

const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos);