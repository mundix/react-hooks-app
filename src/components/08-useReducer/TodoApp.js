import React, { useEffect, useReducer } from 'react';
import { TodoReducer } from './TodoReducer';
import { useForm } from '../../hooks/useForm';

import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

import './styles.css';


const initialState = [{
    id: new Date().getTime(),
    desc: 'Aprender React',
    done: false
}];

const init = () => {
    // Otra cosa que podemos hacer es guardar en el localStorage
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React',
    //     done: false
    // }]
    // ahroa voy a retornar un localSTorage
    return JSON.parse(localStorage.getItem('todos')) || [];
}

//El reducer puede crease en  un archivo independiente 'TodoReducer.js'

export const TodoApp = () => {

    //reducer => es la function reducer, 
    //initialState es el valor inicial
    //el init es una function par ainiciarlizar el state en caso de que el state sea utuilizado , se puede memorizar 
    // const [todos, dispatch] = useReducer(TodoReducer, initialState);
    //Esta funcion init me va ayudar a ejecutar mas rapido, me va ayudar a computar todo y se define const init = () =>
    const [todos, dispatch] = useReducer(TodoReducer, initialState, init);

    // const [formValues, handleInputChange] = useForm({
    // handleInputChange se asigna mediante onChange el input 
    // const [{description}, handleInputChange] = useForm({

    // Esto se va para el TodoApp por que pertenece a esed compoentn (Debio ser removido)
    // const [{ description }, handleInputChange, reset] = useForm({
    //     description: ''
    // });

    // Lio que quiero hacer es grabar ene l localStorage cuando los todo Cambian, puedo usar un useEffect 
    useEffect(() => {
        //si los todos cambian tengo que volver a grabar en el localStorage
        localStorage.setItem('todo', JSON.stringify(todos));
    }, [todos]);

    // 
    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId
        }
        dispatch(action);
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }
    // Esta se encarga de agregar el todo, lo que intersa que cuando se llamae esta function agrege el todo
    // ESte es el quele voy a enviar a mi componetn TodoAdd 
    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    // console.log(formValues);
    // console.log(description);
    // Como hago para borrar el formulario , en el use form crear un metodo reset = () => 

    // Esto tambien se fue para TodoAdd para manejarlo desde alla
    // const handleSubmit = (e) => {
    //     e.preventDefault();

    //     if (description.trim().length <= 1) {
    //         return;
    //     }

    //     const newTodo = {
    //         id: new Date().getTime(),
    //         desc: description,
    //         // desc: 'Nueva Tarea',
    //         done: false
    //     };

    //     const action = {
    //         type: 'add',
    //         payload: newTodo
    //     }

    //     // En el dispatch del useReducer es que se le manda la accion
    //     dispatch(action);
    //     // Y llamo el reset del useForm 
    //     reset();
    // }


    return (
        <>
            <h1>TodoApp ({todos.length})</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    {/* TodoList , argumentos: todos, handleDelete, handleToggle */}
                    <TodoList
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                </div>
                <div className="col-5">
                    <TodoAdd
                        handleAddTodo={handleAddTodo}
                    />
                </div>
            </div>

        </>
    )
}
