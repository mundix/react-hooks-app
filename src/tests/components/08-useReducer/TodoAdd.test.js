import React from 'react';
import { shallow }  from 'enzyme';
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";
import { demoTodo } from '../../fixtures/demoTodo';


describe('Pruebas en <TodoAdd/>', () => {
    
    const handleAddTodo = jest.fn(); 
    const wrapper = shallow(<TodoAdd
        handleAddTodo={handleAddTodo}
    />);


    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('NO debe llamar el handleAddTodo', () => {
        // voy a intentar (simular) disparar el submit del formulario , por consecuencia nod ebe haber ningun value  
        // deberia chocar en la valdiacion de description 
        // puedo hacer referencia ene l form al onSubmit 
        const formSubmit = wrapper.find('form').prop('onSubmit');//DEevuelve Function
        // entonces puedo llamar este formSubmit pero me dara un error si tengo un preventDefault(); 
        // Tengo que enviarlo en el argumento  un objeto { preventDefault(){} }
        formSubmit({preventDefault() {} }); //Esta es una forma de hacerlo. 

        expect(handleAddTodo).toHaveBeenCalledTimes(0); //no se ha llamado ninguna vez

    });

    test('Debe de llamar handleAddTodo', () => {
        const value = 'Aprender React';
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });
        // Esto cambia el input 

        const formSubmit = wrapper.find('form').prop('onSubmit');//DEevuelve Function
        formSubmit({preventDefault() {} }); //Esta es una forma de hacerlo. 

        expect(handleAddTodo).toHaveBeenCalledTimes(1); //no se ha llamado ninguna vez
        // Necesito que se halla llamado con un objeto
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object)); 
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false,
        }); 

        // Borrar el input
        expect(wrapper.find('input').prop('value')).toBe('');
    });
    
    
    
})
