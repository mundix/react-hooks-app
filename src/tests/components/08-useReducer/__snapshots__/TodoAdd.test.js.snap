// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd/> Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-warning mt-1 btn-block"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
